src/rest/rest_utils.c

```diff
@@ -165,100 +557,165 @@ int database_validate_entry(json_t *j_device_object)
-    json_string = json_string_value(j_value);
-
-    device_entry->uuid = strdup(json_string);
-    if (device_entry->uuid == NULL)
+    device_entry = calloc(1, sizeof(database_entry_t));
+    if (device_entry == NULL)
     {
-        return -1;
+        goto exit;
     }
 
-
-    j_value = json_object_get(j_device_object, "psk");
-    json_string = json_string_value(j_value);
-
-    base64_decode(json_string, NULL, &device_entry->psk_len);
-
-    device_entry->psk = (uint8_t *)malloc(device_entry->psk_len);
-    if (device_entry->psk == NULL)
+    mode = database_json_to_entry(j_device_object, "mode", BASE64_NO_ACTION, NULL);
+    if (mode == NULL)
     {
-        return -1;
+        goto exit;
     }
-    base64_decode(json_string, device_entry->psk, &device_entry->psk_len);
-
-
-    j_value = json_object_get(j_device_object, "psk_id");
-    json_string = json_string_value(j_value);
 
-    base64_decode(json_string, NULL, &device_entry->psk_id_len);
+    if (strcasecmp(mode, "psk") == 0)
+    {
+        device_entry->mode = DEVICE_CREDENTIALS_PSK;
+    }
+    else if (strcasecmp(mode, "cert") == 0)
+    {
+        device_entry->mode = DEVICE_CREDENTIALS_CERT;
+    }
+    else if (strcasecmp(mode, "none") == 0)
+    {
+        device_entry->mode = DEVICE_CREDENTIALS_NONE;
+    }
+    else
+    {
+        goto exit;
+    }
 
-    device_entry->psk_id = (uint8_t *)malloc(device_entry->psk_id_len);
-    if (device_entry->psk_id == NULL)
+    device_entry->uuid = database_json_to_entry(j_device_object, "uuid", BASE64_NO_ACTION, NULL);
+    device_entry->name = database_json_to_entry(j_device_object, "name", BASE64_NO_ACTION, NULL);
+    device_entry->public_key = database_json_to_entry(j_device_object, "public_key", BASE64_DECODE, &device_entry->public_key_len);
+    device_entry->secret_key = database_json_to_entry(j_device_object, "secret_key", BASE64_DECODE, &device_entry->secret_key_len);
+    device_entry->serial = database_json_to_entry(j_device_object, "serial", BASE64_DECODE, &device_entry->serial_len);
+
+    if (device_entry->uuid == NULL
+        || device_entry->name == NULL
+        || (device_entry->mode == DEVICE_CREDENTIALS_PSK // some entry types must contain keys that other don't
+        && (device_entry->secret_key == NULL
+        || device_entry->public_key == NULL))
+        || (device_entry->mode == DEVICE_CREDENTIALS_CERT
+        && (device_entry->serial == NULL
+        || device_entry->public_key == NULL)))
     {
-        return -1;
+        goto exit;
     }
-    base64_decode(json_string, device_entry->psk_id, &device_entry->psk_id_len);
 
-    return 0;
+    status = 0;
+exit:
+    if (status)
+    {
+        database_free_entry(device_entry);
+        device_entry = NULL;
+    }
+    return device_entry;
 }
 
-int database_populate_new_entry(json_t *j_new_device_object, database_entry_t *device_entry)
+database_entry_t *database_build_new_entry(json_t *j_device_object, void *context)
```

Comment:
Could you define `context` scope? Does it require whole server context to be passed? Minimal context should be passed to improve simplicity.